<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Reading (and writing) data in R</title>
    <meta charset="utf-8" />
    <meta name="author" content="Marcin Kierczak" />
    <meta name="keywords" content="bioinformatics, course, scilifelab, nbis, R" />
    <script src="site_libs/header-attrs-2.28/header-attrs.js"></script>
    <link href="site_libs/font-awesome-6.4.2/css/all.min.css" rel="stylesheet" />
    <link href="site_libs/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet" />
    <link id="font-awesome-1-attachment" rel="attachment" href="site_libs/font-awesome-6.4.2/fonts/fontawesome-webfont.ttf"/>
    <link rel="stylesheet" href="assets/slide.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Reading (and writing) data in R
]
.subtitle[
## R Foundations for Data Analysis
]
.author[
### Marcin Kierczak
]

---


exclude: true
count: false


&lt;link href="https://fonts.googleapis.com/css?family=Roboto|Source+Sans+Pro:300,400,600|Ubuntu+Mono&amp;amp;subset=latin-ext" rel="stylesheet"&gt;
&lt;link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.3.1/css/all.css" integrity="sha384-mzrmE5qonljUremFsqc01SB46JvROS7bZs3IO2EmfFsd15uHvIt+Y8vEf7N7fWAU" crossorigin="anonymous"&gt;





&lt;!-- ------------ Only edit title, subtitle &amp; author above this ------------ --&gt;





---
name: reading_data

# Reading data

* Can be one of the most consuming and cumbersome aspects of data analysis.

* R provides ways to read and write data stored on different media (e.g.: file, database, url) and in different formats.

* Package `foreign` contains a number of functions to import less common data formats.

---
name: reading_tables

# Reading tables

We can use the `read.table()` function. It is a nice way to read your data into a data frame.  

The function is declared in the following way:  


``` r
read.table(file, header = FALSE, sep = "", quote = "\"'",
           dec = ".", numerals = c("allow.loss", "warn.loss", "no.loss"),
           row.names, col.names, as.is = !stringsAsFactors,
           na.strings = "NA", colClasses = NA, nrows = -1,
           skip = 0, check.names = TRUE, fill = !blank.lines.skip,
           strip.white = FALSE, blank.lines.skip = TRUE,
           comment.char = "#",
           allowEscapes = FALSE, flush = FALSE,
           stringsAsFactors = default.stringsAsFactors(),
           fileEncoding = "", encoding = "unknown", text, skipNul = FALSE)

# or just
read.table(file)
```

---
name: read_table_params

# `read.table()` parameters

You can read all about the `read.table()` function using `?read.table`

The most important arguments are:

* **file** &amp;ndash; the path to the file that contains data, e.g. `/path/to/my/file.csv`
* **header** &amp;ndash; a logical indicating whether the first line of the file contains variable names,
* **sep** &amp;ndash; a character determining variable delimiter, e.g. `","` for csv files,
* **quote** &amp;ndash; a character telling R which character surrounds strings,
* **dec** &amp;ndash; character determining the decimal separator,
* **row/col.names** &amp;ndash; vectors containing row and column names,
* **na.strings** &amp;ndash; a character used for missing data,
* **nrows** &amp;ndash; how many rows should be read,
* **skip** &amp;ndash; how many rows to skip,
* **as.is** &amp;ndash; a vector of logicals or numbers indicating which columns shall not be converted to factors,
* **fill** &amp;ndash; add NA to the end of shorter rows,
* **stringsAsFactors** &amp;ndash; a logical. Rather self explanatory.

---
name: read_table_sibs

# `read.table` and its siblings  

The `read.table` function has some siblings, functions with particular arguments pre-set to a specific value to spare some time:

* `read.csv()` and `read.csv2()` with comma and semicolon as default `sep` and dot and comma as `dec` respectively,
* `read.delim()` and `read.delim2()` for reading tab-delimited files.

We, however, most often use the canonical `read.table()` as it is the most flexible one.

---
name: read_table_example

# `read.table` &amp;mdash; example use


``` r
tab &lt;- read.table(file = 'data/slide_loading_data/2014-04-07_phenos2.csv',
                  sep = ' ',
                  header = T)
tab[1:5, 1:3]
class(tab$reg_no)
```

```
##            id              uid      reg_no
## 1 S11799-2007  S11799/2007_621 S11799/2007
## 2 S14189-2008 S14189/2008_1237 S14189/2008
## 3 S15108-2008 S15108/2008_1224 S15108/2008
## 4 S15299-2007  S15299/2007_515 S15299/2007
## 5 S16860-2009  S16860/2009_849 S16860/2009
## [1] "character"
```

---
name: handling_errors

# What if you encounter errors?

* R documentation `?` and `??`
* Google &amp;ndash; just type R and copy the error you got without your variable names,
* Open the file using a text editor and see if you can spot anything unusual &amp;ndash; 
  * e.g. has the header line the same number of columns as the first line?

--

# Useful terminal commands for debugging (Linux/OsX)

* `cat phenos.txt | awk -F';' '{print NF}'` prints the number of words in each row. `-F';'` says that semicolon is the delimiter,

* `head -n 5 phenos.txt` prints the 5 first lines of the file,

* `tail -n 5 phenos.txt` prints the 5 last lines of the file,

* `head -n 5 phenos.txt | tail -n 2` will print lines 4 and 5...

* `wc -l phenos.txt` will print the number of lines in the file

* `head -n 2 phenos.txt &gt; test.txt` will write the first 2 lines to a new file

--

If it still does not give you a clue &amp;mdash; just try to load first line of the file.  

--

If this still did not help, split the file in two equal-size parts. Check which part gives the error. Split this part into halves and check which 1/4 gives the error... It is faster than you think!

---
name: writing

# Writing with `write.table()`

`read.table()` has its counterpart, the `write.table()` function (as well ass its siblings, like `write.csv()`). You can read more about it in the documentation, let us show some examples:


``` r
vec &lt;- rnorm(10)
write.table(vec, '') # write to screen
write.table(vec, file = 'vector.txt')

# write to the system clipboard, handy!
write.table(vec, 'clipboard', col.names=F, row.names=F)

# or on OsX
clip &lt;- pipe("pbcopy", "w")                       
write.table(vec, file=clip)                               
close(clip)

# To use in a spreadsheet
write.csv(vec, file = 'spreadsheet.csv')
```

---
name: write_big_data

# Writing big data

* HINT: `write.table()` is rather slow on big data &amp;ndash; it checks types for every column and row and does separate formatting to each. If your data consists of only one type of data, convert it to a matrix using `as.matrix` before you write it!  

* You may want to use function `scan()` that reads files as vectors. The content does not have to be in the tabular form. You can also use scan to read data from keyboard: `typed.data &lt;- scan()`

* If data are written as fixed-width fields, use the `read.fwf()` function.

* Also check out the `readLines()` function that enables you to read data from any stream.

---
name: read_xls_matlab

# Read data in xls/xlsx and Matlab


``` r
library(readxl)
data &lt;- read_xlsx('myfile.xlsx')
```


``` r
library(R.matlab)
data &lt;- readMat("mydata.mat")
```

---
name: remote_data

# Working with remote data


``` r
url &lt;- 'https://en.wikipedia.org/wiki/List_of_countries_by_average_wage'
conn &lt;- url(url, 'r')
raw.data &lt;- readLines(conn)
raw.data[1:3]
```

But data is often tabularized...


``` r
library(rvest)
html &lt;- read_html(url)
tables &lt;- html_nodes(html, 'table')
data &lt;- html_table(tables[4])[[1]]
data[1:5, ]
```

---
name: databases

# Working with databases

It is also relatively easy to work with different databases. We will focus on MySQL and present only one example that uses the *RMySQL* package (check also *RODBC* and *RPostgreSQL*).

``` r
library(RMySQL)
db.conn &lt;- dbConnect(MySQL(), user='me',
                     password='qwerty123',
                     dbname='genes',
                     host='127.0.0.237')
query &lt;- dbSendQuery(db.conn, 'SELECT * FROM table7')
data &lt;- fetch(query, n = - 1)
```

---
name: capabilities

# Capabilities

If you are getting some errors, e.g. trying to connect to a url, you may check whether your system (and R) support particular type of file or connection:

``` r
capabilities()
```

```
##        jpeg         png        tiff       tcltk         X11        aqua 
##        TRUE        TRUE        TRUE        TRUE       FALSE       FALSE 
##    http/ftp     sockets      libxml        fifo      cledit       iconv 
##        TRUE        TRUE       FALSE        TRUE       FALSE        TRUE 
##         NLS       Rprof     profmem       cairo         ICU long.double 
##        TRUE        TRUE        TRUE        TRUE        TRUE        TRUE 
##     libcurl 
##        TRUE
```

&lt;!-- --------------------- Do not edit this and below --------------------- --&gt;

---
name: end_slide
class: end-slide, middle
count: false

# See you at the next lecture!


.end-text[
&lt;p class="smaller"&gt;
&lt;span class="small" style="line-height: 1.2;"&gt;Graphics from &lt;/span&gt;&lt;img src="./assets/freepik.jpg" style="max-height:20px; vertical-align:middle;"&gt;&lt;br&gt;
Created: 22-Oct-2024 • &lt;a href="https://www.scilifelab.se/"&gt;SciLifeLab&lt;/a&gt; • &lt;a href="https://nbis.se/"&gt;NBIS&lt;/a&gt; 
&lt;/p&gt;
]


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="assets/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "4:3",
"highlightLanguage": "r",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false,
"slideNumberFormat": "%current%/%total%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
