<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Loading data into R</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="assets/lab.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
#TOC {
  display: none !important;
}
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img src="assets/logo.svg" id="logo" style="height:30px;margin:0;"/></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="home_schedule.html">Schedule</a>
</li>
<li>
  <a href="home_content.html">Content</a>
</li>
<li>
  <a href="home_precourse.html">Precourse</a>
</li>
<li>
  <a href="home_info.html">Info</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Loading data into R</h1>
<h3 class="subtitle">R Programming Foundation for Life Scientists</h3>

</div>


<div id="introduction" class="section level1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>Up until now we have mostly created the object we worked with on the
fly from within R. The most common use-case is however to read in
different data sets that are stored as files, either somewhere on a
server or locally on your computer. In this exercise we will test some
common ways to import data in R and also show to save data from
R. After this exercise you will know how to:</p>
<ul>
<li>Read data from txt files and save the information as a vector, data frame or
a list.</li>
<li>Identify missing data and correctly encode this at import</li>
<li>Check that imported objects are imported correctly</li>
<li>Read data from online resource</li>
<li>Write data to a file</li>
</ul>
<div id="scan" class="section level2">
<h2><span class="header-section-number">1.1</span> <code>scan()</code></h2>
<p>The function <code>scan()</code> can be used both to read data from files and
directly from keyboard. The function is very flexible and have many
different settings that allow to read data in different formats. To
read and store a set of words that you type on your keyboard try the
following code that will prompt your for input. After each word press
enter and R will prompt you for new input. After the last word have
been typed press enter twice to get back to your R prompt and have
your character vector named words available in R your session.</p>
<pre><code>words &lt;- scan(what = character())</code></pre>
<div id="exercises" class="section level3">
<h3><span class="header-section-number">1.1.1</span> Exercises</h3>
<p>Download the file book chapter from this <a href="../files/book_chapter.txt">link</a>. Read the manual for
scan and read the text file named book_chapter.txt into R, first as
vector and then as a list, with each word in the chapter saved as a
entry in the vector or as a single vector in a list.</p>
<details>
<summary>:key: Click to see how</summary>
<pre>
    shelley.vec <- scan(file = "book_chapter.txt", what = character())
    str(shelley.vec)

    shelley.list <- scan(file = "book_chapter.txt", what = list(character()))
    class(shelley.list)

    Read 420 items

    chr [1:420] "My" "present" "situation" "was" "one" "in" ...
    Read 420 records
    [1] "list"
</pre>
</details>
<p><br></p>
<p>Check that your newly created objects contain the correct information
and have been saved as you have intended eg. each entry of the vector
or the list should contain a single word. Once your convinced that you
have a sound word vector and list.</p>
<ol style="list-style-type: decimal">
<li><p>Identify the longest word in your vector.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    sort(nchar(shelley.vec), decreasing = TRUE)
    which(nchar(shelley.vec) == max(nchar(shelley.vec)))
    shelley.vec[381]

    [1] 690  12  11  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10
    [19]   9   9   9   9   9   9   9   9   9   9   9   9   9   9   9   9   9   9
    [37]   9   9   8   8   8   8   8   8   8   8   8   8   8   8   8   8   8   8
    [55]   8   8   8   8   8   8   8   8   8   8   7   7   7   7   7   7   7   7
    [73]   7   7   7   7   7   7   7   7   7   7   7   7   7   7   7   7   7   6
    [91]   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6
    [109]   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6   6
    [127]   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5
    [145]   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5
    [163]   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5   5
    [181]   5   5   5   5   5   4   4   4   4   4   4   4   4   4   4   4   4   4
    [199]   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4
    [217]   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4   4
    [235]   4   4   4   4   4   4   4   3   3   3   3   3   3   3   3   3   3   3
    [253]   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3
    [271]   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3
    [289]   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3
    [307]   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3   3
    [325]   3   3   3   3   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    [343]   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    [361]   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    [379]   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2   2
    [397]   2   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1   1
    [415]   1   1   1   1   1   1
    [1] 381
    [1] "By the sacred earth on which I kneel, by the shades that wander near me, by the deep and eternal grief that I feel, I swear; and by thee, O Night, and the spirits that preside over thee, to pursue the daemon who caused this misery, until he or I shall perish in mortal conflict. For this purpose I will preserve my life; to execute this dear revenge will I again behold the sun and tread the green herbage of earth, which otherwise should vanish from my eyes forever. And I call on you, spirits of the dead, and on you, wandering ministers of vengeance, to aid and conduct me in my work. Let the cursed and hellish monster drink deep of agony; let him feel the despair that now torments me."
</pre>
</details>
<p><br></p></li>
<li><p>Go back and fix the way you read in the text to make sure that you
get a vector with all words in chapter as individual entries also
filter any non-letter characters and now identify the longest word.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    shelley.vec2 <- scan(file = "book_chapter.txt", what = 'character', sep = ' ', quote = NULL)
    shelley.filt2 <- gsub(pattern = '[^[:alnum:] ]', replacement = "", x = shelley.vec2)
    which(nchar(shelley.filt2) == max(nchar(shelley.filt2)))
    shelley.filt2[301]

    Read 551 items
    [1] 301
    [1] "uninterested"
</pre>
</details>
<p><br></p></li>
</ol>
</div>
</div>
<div id="the-read.table-function" class="section level2">
<h2><span class="header-section-number">1.2</span> The read.table function<a id="orgheadline3"></a></h2>
<p>This is the by far most common way to get data into R. As the function
creates a data frame at import it will only work for data set that
fits those criteria, meaning that the data needs to have a set of
columns of equal length that are separated with a common string
eg. tab, comma, semicolon etc.</p>
<p>In this code block with first import the data from normalized.txt from
a file (that you can get <a href="../files/normalized.txt">here</a>) and accept the defaults for all other
arguments in the function. With this settings R will read it as a tab
delimited file and will use the first row of the data as colnames
(header) and the first column as rownames.</p>
<pre><code>expr.At &lt;- read.table(&quot;normalized.txt&quot;)
head(expr.At)

              bZIP29_1  bZIP29_2  bZIP29_3        WT_1      WT_2       WT_3
AT1G01020.1  13.572739 14.167143 12.972703  14.8181738 15.904017 11.3270623
AT1G01030.1   1.417234  1.201454  1.385434   0.8590246  1.096829  0.8596431
AT1G01040.2  41.862906 41.199853 42.696566  37.5286358 34.849241 48.6456871
AT1G01050.1 102.422397 98.318969 92.068406 104.2104178 97.418336 86.0654463
AT1G01060.1   3.216031  4.004846  3.589534   3.7045434  2.642360  6.2703380
AT1G01070.1   8.230858 17.871625 14.924906   7.9996663  8.824486 13.8554244</code></pre>
<p>One does however not have to have all data as a file an the local
disk, instead one can read data from online resources. The following
command will read in a file from a web server.</p>
<pre><code>url = &#39;http://archive.ics.uci.edu/ml/machine-learning-databases/abalone/abalone.data&#39;
abalone = read.table(url, header = F , sep = &#39;,&#39;)
head(abalone)

  V1    V2    V3    V4     V5     V6     V7    V8 V9
1  M 0.455 0.365 0.095 0.5140 0.2245 0.1010 0.150 15
2  M 0.350 0.265 0.090 0.2255 0.0995 0.0485 0.070  7
3  F 0.530 0.420 0.135 0.6770 0.2565 0.1415 0.210  9
4  M 0.440 0.365 0.125 0.5160 0.2155 0.1140 0.155 10
5  I 0.330 0.255 0.080 0.2050 0.0895 0.0395 0.055  7
6  I 0.425 0.300 0.095 0.3515 0.1410 0.0775 0.120  8</code></pre>
<div id="exercises-1" class="section level3">
<h3><span class="header-section-number">1.2.1</span> Exercises</h3>
<ol style="list-style-type: decimal">
<li><p>Download the file <a href="../files/example.data">example.data</a> to your
computer and import it to R using the read.table function. This
files consist of gene expression values. Once you have the object
in R validate that it looks okay and export it using the
write.table function.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>
    ed <- read.table("example.data", sep = ":")
    head(ed)
    str(ed)
</pre>
</details>
<p><br>
Encode all NA values as “missing”, at export.<br />
</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>
write.table(x = ed, na = "missing", file = "example_mis.data")
</pre>
</details>
<p><br></p></li>
<li><p>Read in the file you just created and double-check that you have the same data as earlier.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>
    df.test <- read.table("example_mis.data", na.strings = "missing")
</pre>
</details>
<p><br></p></li>
<li><p>Analysing genome annotation in R using read.table</p></li>
</ol>
<p>For this exercise we will load a GTF file into R and calculate some
basic summary statistics from the file. In the first part we will use
basic manipulations of data frames to extract the information. In the
second part you get a try out a library designed to work with
annotation data, that stores the information in a more complex format,
that allow for easy manipulation and calculation of summaries from
genome annotation files.</p>
<p>For those not familiar with the gtf format it is a file format
containing annotation information for a genome. It does not contain
the actual DNA sequence of the organism, but instead refers to
positions along the genome.</p>
<p>A valid GTF file should contain the following tab delimited fields
(taken from the ensembl home page).</p>
<ol style="list-style-type: decimal">
<li>seqname - name of the chromosome or scaffold; chromosome names can
be given with or without the ‘chr’ prefix.</li>
<li>source - name of the program that generated this feature, or the
data source (database or project name)</li>
<li>feature - feature type name, e.g. Gene, Variation, Similarity</li>
<li>start - Start position of the feature, with sequence numbering
starting at 1.</li>
<li>end - End position of the feature, with sequence numbering starting
at 1.</li>
<li>score - A floating point value.</li>
<li>strand - defined as + (forward) or - (reverse).</li>
<li>frame - One of ‘0’, ‘1’ or ‘2’. ‘0’ indicates that the first base
of the feature is the first base of a codon, ‘1’ that the second
base is the first base of a codon, and so on..</li>
<li>attribute - A semicolon-separated list of tag-value pairs,
providing additional information about each feature.</li>
</ol>
<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">
<colgroup>
<col  class="org-right" />
<col  class="org-left" />
<col  class="org-left" />
<col  class="org-right" />
<col  class="org-right" />
<col  class="org-left" />
<col  class="org-left" />
<col  class="org-left" />
<col  class="org-left" />
</colgroup>
<thead>
<tr>
<th scope="col" class="org-right">
1.
</th>
<th scope="col" class="org-left">
2.
</th>
<th scope="col" class="org-left">
3.
</th>
<th scope="col" class="org-right">
4.
</th>
<th scope="col" class="org-right">
5.
</th>
<th scope="col" class="org-left">
6.
</th>
<th scope="col" class="org-left">
7.
</th>
<th scope="col" class="org-left">
8.
</th>
<th scope="col" class="org-left">
9.
</th>
</tr>
</thead>
<tbody>
<tr>
<td class="org-right">
1
</td>
<td class="org-left">
transcribed_unprocessed_pseudogene
</td>
<td class="org-left">
gene
</td>
<td class="org-right">
11869
</td>
<td class="org-right">
14409
</td>
<td class="org-left">
.
</td>
<td class="org-left">
+
</td>
<td class="org-left">
.
</td>
<td class="org-left">
gene_id; “ENSG00000223972”;
</td>
</tr>
<tr>
<td class="org-right">
1
</td>
<td class="org-left">
processed_transcript
</td>
<td class="org-left">
transcript
</td>
<td class="org-right">
11869
</td>
<td class="org-right">
14409
</td>
<td class="org-left">
.
</td>
<td class="org-left">
+
</td>
<td class="org-left">
.
</td>
<td class="org-left">
gene_id; “ENSG00000223972”;
</td>
</tr>
</tbody>
</table>
<p>The last column can contain a large number of attributes that are
comma-separated.</p>
<p>As these files for many organisms are large we will in this exercise
use the latest version of Drosophila melanogaster genome annotation
available at
<a href="ftp://ftp.ensembl.org/pub/release-86/gtf/drosophila_melanogaster" class="uri">ftp://ftp.ensembl.org/pub/release-86/gtf/drosophila_melanogaster</a> that
is small enough for analysis even on a laptop.</p>
<p>Open this and download the file named
Drosophila_melanogaster.BDGP6.86.gtf.gz to your computer. Unzip this
file and keep track of where your store the file.</p>
With this done read this file into R using the function read.table and
add meaningful column names to the table.
<details>
<summary>:key: Click to see how</summary>
<pre>

    d.gtf <- read.table("Drosophila_melanogaster.BDGP6.86.gtf",
                        header = FALSE, comment.char = "#", sep = "\t")
    colnames(d.gtf) <- c("Chromosome", "Source", "Feature", "Start",
                         "End", "Score", "Strand", "Frame", "Attribute")
</pre>
</details>
<p><br></p>
<p>Prior to any analysis you should make sure that your attempt to read
in the file has worked as expected. This can for example be done by
having a look at the dimension of the stored object and making sure
that it has the structure you expect.</p>
<details>
<summary>:key: Click to see how</summary>
<pre>

    dim(d.gtf)
    str(d.gtf)

    [1] 538684      9
    'data.frame':   538684 obs. of  9 variables:
     $ Chromosome: Factor w/ 57 levels "211000022278158",..: 51 51 51 51 51 51 51 51 51 51 ...
     $ Source    : Factor w/ 2 levels "FlyBase","ensembl": 1 1 1 1 1 1 1 1 1 1 ...
     $ Feature   : Factor w/ 9 levels "CDS","Selenocysteine",..: 5 9 3 5 9 3 1 6 3 1 ...
     $ Start     : int  722370 722370 722370 835381 835381 835381 835381 835381 869486 869486 ...
     $ End       : int  722621 722621 722621 2503907 2503907 835491 835491 835383 869548 869548 ...
     $ Score     : Factor w/ 1 level ".": 1 1 1 1 1 1 1 1 1 1 ...
     $ Strand    : Factor w/ 2 levels "+","-": 2 2 2 1 1 1 1 1 1 1 ...
     $ Frame     : Factor w/ 4 levels ".","0","1","2": 1 1 1 1 1 1 2 2 1 2 ...
     $ Attribute : Factor w/ 455338 levels "gene_id FBgn0000003; gene_name 7SLRNA:CR32864; gene_source FlyBase; gene_biotype lincRNA;",..: 348579 348581 348580 453172 453215 453194 453195 453193 453199 453200 ...
</pre>
</details>
<p><br></p>
<ol style="list-style-type: decimal">
<li><p>How many chromosome names can be found in the annotation file?</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    levels(d.gtf$Chromosome)

     [1] "211000022278158"           "211000022278279"          
     [3] "211000022278282"           "211000022278298"          
     [5] "211000022278307"           "211000022278309"          
     [7] "211000022278436"           "211000022278449"          
     [9] "211000022278498"           "211000022278522"          
    [11] "211000022278603"           "211000022278604"          
    [13] "211000022278664"           "211000022278724"          
    [15] "211000022278750"           "211000022278760"          
    [17] "211000022278875"           "211000022278877"          
    [19] "211000022278878"           "211000022278879"          
    [21] "211000022278880"           "211000022278985"          
    [23] "211000022279055"           "211000022279108"          
    [25] "211000022279132"           "211000022279134"          
    [27] "211000022279165"           "211000022279188"          
    [29] "211000022279222"           "211000022279264"          
    [31] "211000022279342"           "211000022279392"          
    [33] "211000022279446"           "211000022279528"          
    [35] "211000022279529"           "211000022279531"          
    [37] "211000022279555"           "211000022279681"          
    [39] "211000022279708"           "211000022280133"          
    [41] "211000022280328"           "211000022280341"          
    [43] "211000022280347"           "211000022280481"          
    [45] "211000022280494"           "211000022280645"          
    [47] "211000022280703"           "2L"                       
    [49] "2R"                        "3L"                       
    [51] "3R"                        "4"                        
    [53] "Unmapped_Scaffold_8"       "X"                        
    [55] "Y"                         "dmel_mitochondrion_genome"
    [57] "rDNA"

</pre>
</details>
<p><br></p></li>
<li><p>How many exons is there in total and per chromosome?</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    aggregate(d.gtf$Feature, by = list(d.gtf$Chromosome), summary)

                         Group.1 x.CDS x.Selenocysteine x.exon x.five_prime_utr
    1            211000022278158     0                0      1                0
    2            211000022278279     0                0      1                0
    3            211000022278282     0                0      1                0
    4            211000022278298     0                0      1                0
    5            211000022278307     0                0      1                0
    6            211000022278309     0                0      1                0
    7            211000022278436     0                0      1                0
    8            211000022278449     0                0      2                0
    9            211000022278498     0                0      1                0
    10           211000022278522     0                0      1                0
    11           211000022278603     0                0      1                0
    12           211000022278604     0                0      1                0
    13           211000022278664     0                0      1                0
    14           211000022278724     0                0      1                0
    15           211000022278750     0                0      1                0
    16           211000022278760     2                0      2                0
    17           211000022278875     0                0      1                0
    18           211000022278877     0                0      1                0
    19           211000022278878     0                0      1                0
    20           211000022278879     0                0      1                0
    21           211000022278880     0                0      1                0
    22           211000022278985     0                0      1                0
    23           211000022279055     0                0      1                0
    24           211000022279108     0                0      1                0
    25           211000022279132     0                0      1                0
    26           211000022279134     0                0      1                0
    27           211000022279165     0                0      1                0
    28           211000022279188     3                0      3                1
    29           211000022279222     0                0      1                0
    30           211000022279264     0                0      1                0
    31           211000022279342     0                0      1                0
    32           211000022279392     0                0      1                0
    33           211000022279446     0                0      1                0
    34           211000022279528     0                0      1                0
    35           211000022279529     0                0      1                0
    36           211000022279531     0                0      1                0
    37           211000022279555     0                0      1                0
    38           211000022279681     0                0      1                0
    39           211000022279708     0                0      1                0
    40           211000022280133     0                0      1                0
    41           211000022280328     4                0      4                1
    42           211000022280341     0                0      1                0
    43           211000022280347     0                0      1                0
    44           211000022280481     0                0      1                0
    45           211000022280494     0                0      2                0
    46           211000022280645     0                0      1                0
    47           211000022280703     0                0      1                0
    48                        2L 28047                2  32747             8374
    49                        2R 33657                0  38551             8842
    50                        3L 29473                0  34347             8738
    51                        3R 37167                0  43159            10693
    52                         4  2732                0   3165              570
    53       Unmapped_Scaffold_8    12                0     14                4
    54                         X 28967                2  34136             8824
    55                         Y   111                0    182               13
    56 dmel_mitochondrion_genome    13                0     37                0
    57                      rDNA     0                0     21                0
</pre>
</details>
<p><br></p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    by(data = d.gtf$Feature, d.gtf[,"Chromosome"], summary)
    d.gtf[, "Chromosome"]: 211000022278158
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278279
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278282
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278298
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278307
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278309
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278436
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278449
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               2               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278498
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278522
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278603
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278604
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278664
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278724
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278750
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278760
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  2               0               2               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  1               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278875
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278877
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
        d.gtf[, "Chromosome"]: 211000022278878
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278879
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278880
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022278985
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279055
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279108
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279132
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279134
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279165
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279188
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  3               0               3               1               1
        start_codon      stop_codon three_prime_utr      transcript
                  1               1               1               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279222
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279264
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279342
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279392
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279446
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279528
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279529
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279531
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279555
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279681
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022279708
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280133
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280328
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  4               0               4               1               1
        start_codon      stop_codon three_prime_utr      transcript
                  1               1               1               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280341
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280347
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280481
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280494
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               2               0               2
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               2
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280645
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 211000022280703
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0               1               0               1
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0               1
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 2L
                CDS  Selenocysteine            exon  five_prime_utr            gene
              28047               2           32747            8374            3465
        start_codon      stop_codon three_prime_utr      transcript
               5675            5656            6142            6632
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 2R
                CDS  Selenocysteine            exon  five_prime_utr            gene
              33657               0           38551            8842            3601
        start_codon      stop_codon three_prime_utr      transcript
               6028            6018            6484            6927
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 3L
                CDS  Selenocysteine            exon  five_prime_utr            gene
              29473               0           34347            8738            3433
        start_codon      stop_codon three_prime_utr      transcript
               5873            5853            6352            6676
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 3R
                CDS  Selenocysteine            exon  five_prime_utr            gene
              37167               0           43159           10693            4125
        start_codon      stop_codon three_prime_utr      transcript
               7105            7092            7782            8010
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: 4
                CDS  Selenocysteine            exon  five_prime_utr            gene
               2732               0            3165             570             111
        start_codon      stop_codon three_prime_utr      transcript
                295             289             339             343
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: Unmapped_Scaffold_8
                CDS  Selenocysteine            exon  five_prime_utr            gene
                 12               0              14               4               2
        start_codon      stop_codon three_prime_utr      transcript
                  3               3               2               3
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: X
                CDS  Selenocysteine            exon  five_prime_utr            gene
              28967               2           34136            8824            2647
        start_codon      stop_codon three_prime_utr      transcript
               5372            5351            5921            5973
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: Y
                CDS  Selenocysteine            exon  five_prime_utr            gene
                111               0             182              13              71
        start_codon      stop_codon three_prime_utr      transcript
                 23              22              10              72
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: dmel_mitochondrion_genome
                CDS  Selenocysteine            exon  five_prime_utr            gene
                 13               0              37               0              37
        start_codon      stop_codon three_prime_utr      transcript
                 12              10               0              37
    ------------------------------------------------------------
    d.gtf[, "Chromosome"]: rDNA
                CDS  Selenocysteine            exon  five_prime_utr            gene
                  0               0              21               0              19
        start_codon      stop_codon three_prime_utr      transcript
                  0               0               0              19
</pre>
</details>
<p><br></p></li>
<li><p>Filter the data frame to only retain gene annotations</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>
    d.gtf.gene <- d.gtf[d.gtf$Feature == "gene",]
</pre>
</details>
<p><br></p></li>
<li><p>What is the average gene length of in the Drosophila genome?</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

mean(abs(d.gtf.gene$Start - d.gtf.gene$End))

[1] 5753.282
</pre>
</details>
<p><br></p></li>
<li><p>What fraction of the genes are encoded on the plus strand of
the genome.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

sum(d.gtf.gene$Strand == "+") / length(d.gtf.gene$Strand)

[1] 0.5016231
</pre>
</details>
<p><br></p></li>
<li><p>What is the median and mean length of the exons found on chromosome
3R in the data set?</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

    d.gtf3R <- d.gtf[d.gtf$Chromosome == "3R",]
    exon.position <- d.gtf3R[d.gtf3R$Feature == "exon",c("Start", "End")]   
    median(abs(exon.position$Start - exon.position$End))
    mean(abs(exon.position$Start - exon.position$End))

    [1] 251
    [1] 468.6693
</pre>
</details>
<p><br></p></li>
<li><p>Do the same calculations for the chromosomes 2L, 2R, 3L, 4, X and Y
using a for loop.</p>
<details>
<p><summary>:key: Click to see how</summary></p>
<pre>

         chr <- c("2L", "2R", "3L", "4", "X", "Y")
         for (i in chr) {
              d.gtf.tmp <- d.gtf[d.gtf$Chromosome == i,]
              exon.position <- d.gtf.tmp[d.gtf.tmp$Feature == "exon", c("Start", "End")]   
              exon.med <- median(abs(exon.position$Start - exon.position$End))
              exon.mean <- mean(abs(exon.position$Start - exon.position$End))
              txt <- sprintf("The median and mean exon length for %s is %g and %g, respectively", i, exon.med, exon.mean)
              print(txt)
    }

    [1] "The median and mean exon length for 2L is 279 and 502.617, respectively"
    [1] "The median and mean exon length for 2R is 225 and 437.187, respectively"
    [1] "The median and mean exon length for 3L is 255 and 502.116, respectively"
    [1] "The median and mean exon length for 4 is 198 and 429.573, respectively"
    [1] "The median and mean exon length for X is 257 and 526.301, respectively"
    [1] "The median and mean exon length for Y is 410.5 and 657.055, respectively"
</pre>
</details>
<p><br></p></li>
</ol>
</div>
</div>
</div>


<footer class="footer">
<div class="container footer-container">
<div class="row">
  
<div class="col-sm-8" style="text-align:left;vertical-align:middle;">
<p class="small" style="color:#bdbdbd;">
<b>2020</b> • NBIS • SciLifeLab
</p>
</div>

<div class="col-sm-4" style="text-align:right;vertical-align:middle;">
<p style="color:#bdbdbd;">
<span>
  <span class="footericon" style="padding-right:4px; padding-left:4px">
    <a href="https://nbis.se/">
      <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 496 512"><path d="M248 8C111.03 8 0 119.03 0 256s111.03 248 248 248 248-111.03 248-248S384.97 8 248 8zm82.29 357.6c-3.9 3.88-7.99 7.95-11.31 11.28-2.99 3-5.1 6.7-6.17 10.71-1.51 5.66-2.73 11.38-4.77 16.87l-17.39 46.85c-13.76 3-28 4.69-42.65 4.69v-27.38c1.69-12.62-7.64-36.26-22.63-51.25-6-6-9.37-14.14-9.37-22.63v-32.01c0-11.64-6.27-22.34-16.46-27.97-14.37-7.95-34.81-19.06-48.81-26.11-11.48-5.78-22.1-13.14-31.65-21.75l-.8-.72a114.792 114.792 0 0 1-18.06-20.74c-9.38-13.77-24.66-36.42-34.59-51.14 20.47-45.5 57.36-82.04 103.2-101.89l24.01 12.01C203.48 89.74 216 82.01 216 70.11v-11.3c7.99-1.29 16.12-2.11 24.39-2.42l28.3 28.3c6.25 6.25 6.25 16.38 0 22.63L264 112l-10.34 10.34c-3.12 3.12-3.12 8.19 0 11.31l4.69 4.69c3.12 3.12 3.12 8.19 0 11.31l-8 8a8.008 8.008 0 0 1-5.66 2.34h-8.99c-2.08 0-4.08.81-5.58 2.27l-9.92 9.65a8.008 8.008 0 0 0-1.58 9.31l15.59 31.19c2.66 5.32-1.21 11.58-7.15 11.58h-5.64c-1.93 0-3.79-.7-5.24-1.96l-9.28-8.06a16.017 16.017 0 0 0-15.55-3.1l-31.17 10.39a11.95 11.95 0 0 0-8.17 11.34c0 4.53 2.56 8.66 6.61 10.69l11.08 5.54c9.41 4.71 19.79 7.16 30.31 7.16s22.59 27.29 32 32h66.75c8.49 0 16.62 3.37 22.63 9.37l13.69 13.69a30.503 30.503 0 0 1 8.93 21.57 46.536 46.536 0 0 1-13.72 32.98zM417 274.25c-5.79-1.45-10.84-5-14.15-9.97l-17.98-26.97a23.97 23.97 0 0 1 0-26.62l19.59-29.38c2.32-3.47 5.5-6.29 9.24-8.15l12.98-6.49C440.2 193.59 448 223.87 448 256c0 8.67-.74 17.16-1.82 25.54L417 274.25z"></path>
      </svg>
    </a>
    </span>
    <span class="footericon" style="padding-right:4px; padding-left:4px">
      <a href="https://twitter.com/NBISwe"><svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 512 512"><path d="M459.37 151.716c.325 4.548.325 9.097.325 13.645 0 138.72-105.583 298.558-298.558 298.558-59.452 0-114.68-17.219-161.137-47.106 8.447.974 16.568 1.299 25.34 1.299 49.055 0 94.213-16.568 130.274-44.832-46.132-.975-84.792-31.188-98.112-72.772 6.498.974 12.995 1.624 19.818 1.624 9.421 0 18.843-1.3 27.614-3.573-48.081-9.747-84.143-51.98-84.143-102.985v-1.299c13.969 7.797 30.214 12.67 47.431 13.319-28.264-18.843-46.781-51.005-46.781-87.391 0-19.492 5.197-37.36 14.294-52.954 51.655 63.675 129.3 105.258 216.365 109.807-1.624-7.797-2.599-15.918-2.599-24.04 0-57.828 46.782-104.934 104.934-104.934 30.213 0 57.502 12.67 76.67 33.137 23.715-4.548 46.456-13.32 66.599-25.34-7.798 24.366-24.366 44.833-46.132 57.827 21.117-2.273 41.584-8.122 60.426-16.243-14.292 20.791-32.161 39.308-52.628 54.253z"></path>
    </svg>
    </a>
    </span>
    <span class="footericon" style="padding-left:4px">
      <a href="https://www.linkedin.com/company/nbisweden/">
        <svg style="height:0.8em;top:.04em;position:relative;fill:#bdbdbd;" viewBox="0 0 448 512"><path d="M416 32H31.9C14.3 32 0 46.5 0 64.3v383.4C0 465.5 14.3 480 31.9 480H416c17.6 0 32-14.5 32-32.3V64.3c0-17.8-14.4-32.3-32-32.3zM135.4 416H69V202.2h66.5V416zm-33.2-243c-21.3 0-38.5-17.3-38.5-38.5S80.9 96 102.2 96c21.2 0 38.5 17.3 38.5 38.5 0 21.3-17.2 38.5-38.5 38.5zm282.1 243h-66.4V312c0-24.8-.5-56.7-34.5-56.7-34.6 0-39.9 27-39.9 54.9V416h-66.4V202.2h63.7v29.2h.9c8.9-16.8 30.6-34.5 62.9-34.5 67.2 0 79.7 44.3 79.7 101.9V416z"></path>
        </svg>
    </a>
    </span>
  </span>
</p>
</div>

</div>
</div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
